cli/src/pcluster/models/common.py

```diff
@@ -239,3 +252,413 @@ def dump_json(self):
+                body=yaml.dump(content),
+                key=self.get_object_key(file_type, file_name),
+            )
+        elif format == "json":
+            result = AWSApi.instance().s3.put_object(
+                bucket_name=self.name,
+                body=json.dumps(content),
+                key=self.get_object_key(file_type, file_name),
+            )
+        else:
+            result = AWSApi.instance().s3.put_object(
+                bucket_name=self.name,
+                body=content,
+                key=self.get_object_key(file_type, file_name),
+            )
+        return result
+
+    def _get_file(self, file_name, file_type, version_id=None):
+        """Get file from S3 bucket."""
+        result = AWSApi.instance().s3.get_object(
+            bucket_name=self.name, key=self.get_object_key(file_type, file_name), version_id=version_id
+        )
+
+        config_content = result["Body"].read().decode("utf-8")
+        return yaml.safe_load(config_content)
+
+    def _get_file_url(self, file_name, file_type):
+        """Get file url from S3 bucket."""
+        url = "https://{bucket_name}.s3.{region}.amazonaws.com{partition_suffix}/{config_key}".format(
+            bucket_name=self.name,
+            region=self.region,
+            partition_suffix=".cn" if self.region.startswith("cn") else "",
+            config_key=self.get_object_key(file_type, file_name),
+        )
+        return url
+
+
+class S3BucketFactory:
+    """S3 bucket factory to return a bucket object with existence check and creation."""
+
+    @staticmethod
+    def init_s3_bucket(bucket_type, service_name, stack_name, custom_s3_bucket):
+        """Initialize the s3 bucket."""
+        if custom_s3_bucket:
+            bucket = S3Bucket(
+                service_name=service_name,
+                name=custom_s3_bucket,
+                bucket_type=bucket_type,
+                stack_name=stack_name,
+                is_custom_bucket=True,
+            )
+            try:
+                bucket.check_bucket_exists()
+            except AWSClientError as e:
+                raise AWSClientError(
+                    "check_bucket_exists", f"Unable to access config-specified S3 bucket {bucket.name}. Due to {str(e)}"
+                )
+        else:
+            bucket = S3Bucket(service_name=service_name, bucket_type=bucket_type, stack_name=stack_name)
+            try:
+                try:
+                    bucket.check_bucket_exists()
+                except AWSClientError as e:
+                    if "Not Found" in str(e):
+                        bucket.create_bucket()
+                    else:
+                        raise
+            except Exception as e:
+                LOGGER.error("Unable to create S3 bucket %s.", bucket.name)
+                raise AWSClientError("create_bucket", str(e))
+        return bucket
+
+    @staticmethod
+    def get_cluster_bucket(service_name, stack_name, bucket_name):
```

Comment:
I'm not sure I understand why we need these extra get_*_bucket methods. I would avoid the dependency on the S3BucketFactory from the ClusterStack. isn't the bucket already part of the Cluster object?

I believe the error is that we are misusing the ClusterStack class (not as part of this PR) and this is causing this unneeded coupling. ClusterStack should not contain the logic to fetch the cluster configuration nor the logic to handle the keep logs on deletion but needs to simply reflect the CloudFormation stack status with a series of properties to retrieve the stack info and outputs. On top of that it can expose simple create, update  and delete operations.
The retrieval of the config from the bucket and the logic to preserve logs need to be moved into the Cluster class.
I know it's not directly part of this PR but I'm afraid that the best path forward is to fix that and then apply your changes.

Reply:
remove